frontend/__tests__/pages/index.test.js

```diff
@@ -0,0 +1,143 @@
+    render(<Index />);
+
+    expect(screen.getByText("My Library")).toBeInTheDocument();
+    expect(screen.getByText("DU Capstone")).toBeInTheDocument();
+    expect(screen.getByText(booksData[0].title)).toBeInTheDocument();
+    expect(
+      screen.getByText(
+        `${booksData[0].author.firstName} ${booksData[0].author.lastName}`
+      )
+    ).toBeInTheDocument();
+    expect(screen.getByText(booksData[1].title)).toBeInTheDocument();
+    expect(
+      screen.getByText(
+        `${booksData[1].author.firstName} ${booksData[1].author.lastName}`
+      )
+    ).toBeInTheDocument();
+  });
+  it("should close Add Book Modal if onClose is called", () => {
+    useGetAuthors.mockReturnValue({});
+    useGetBooks.mockReturnValue({ books: [] });
+    useAddAuthor.mockReturnValue({ addAuthor: () => {} });
+    useAddBook.mockReturnValue({ addBook: () => {} });
+    AddBookModal.mockImplementation(({ onClose }) => {
+      return <button onClick={onClose}>Book Is Being Added</button>;
+    });
+    render(<Index />);
+
+    userEvent.click(screen.getByText("+ Add Book"));
+    expect(screen.getByText("Book Is Being Added")).toBeInTheDocument();
+
+    userEvent.click(screen.getByText("Book Is Being Added"));
+    expect(screen.queryByText("Book Is Being Added")).toBeNull();
+  });
+  it("should add book if onSubmit is called", async () => {
+    const authorData = {
+      id: "1",
+      firstName: "Encyclopedia",
+      lastName: "Brown",
+    };
+    const refetchAuthorsMock = jest.fn(() => null);
+    const addAuthorMock = jest.fn(() => authorData);
+    const addBookMock = jest.fn();
+    useGetAuthors.mockReturnValue({ refetchAuthors: refetchAuthorsMock });
+    useGetBooks.mockReturnValue({ books: [] });
+    useAddAuthor.mockReturnValue({ addAuthor: addAuthorMock });
+    useAddBook.mockReturnValue({ addBook: addBookMock });
+    const bookInputData = {
+      title: "Book",
+      author: "Encyclopedia Brown",
+      description:
+        'Encyclopedia Brown is a series of books featuring the adventures of boy detective Leroy Brown, nicknamed "Encyclopedia" for his intelligence and range of knowledge. The series of 29 children\'s novels was written (one co-written) by Donald J. Sobol, with the first book published in 1963 and the last published posthumously in 2012.',
+    };
+    AddBookModal.mockImplementation(({ onSubmit }) => {
+      return (
+        <button onClick={() => onSubmit(bookInputData)}>
+          Book Is Being Added
+        </button>
+      );
+    });
+    render(
+      <ToastProvider>
+        <Index />
+      </ToastProvider>
+    );
+
+    userEvent.click(screen.getByText("+ Add Book"));
+    expect(screen.getByText("Book Is Being Added")).toBeInTheDocument();
+
+    userEvent.click(screen.getByText("Book Is Being Added"));
+    expect(screen.queryByText("Book Is Being Added")).toBeNull();
+    await waitFor(() => screen.findByText("Book was added to the library"), {
+      timeout: 3000,
+    });
```

Comment:
Please explain the approach used here

Reply:
The state that controls whether the AddBookModal is rendered is initially set to False so we do not see the mocked component, which is a button with the text "Book was added to the library." Once I click the "Add Book" button, that state switches to True. In order to test that part of the code, I needed the jest runner to wait for the mocked component to show up. React testing library provides waitFor as a way to make the jest runner wait until either the mocked component is found or some time passes which results in a timeout.